libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/MapboxNavigationObserver.kt

```diff
@@ -0,0 +1,65 @@
+package com.mapbox.navigation.ui.lifecycle
+
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.MapboxNavigation
+
+/**
+ * This allows you to attach and detach [MapboxNavigation] observers
+ * from any component with a lifecycle.
+ *
+ * Example of observing locations with a view model
+ * ```
+ * class MyViewModel : ViewModel() {
+ *   private val locationObserver = LocationObserver()
+ *
+ *   val location: LiveData<Location> = locationObserver.location.asLiveData()
+ *
+ *   init {
+ *     MapboxNavigationApp.register(myMapboxNavigationObserver)
+ *   }
+ *
+ *   override fun onCleared() {
+ *     MapboxNavigationApp.unregister(myMapboxNavigationObserver)
+ *   }
+ * }
+ *
+ * class LocationNavigationObserver : MapboxNavigationObserver {
+ *   private val mutableLocation = MutableStateFlow<LocationMatcherResult?>(null)
+ *   val locationFlow: Flow<LocationMatcherResult?> = mutableLocation
+ *
+ *   private val locationObserver = object : LocationObserver {
+ *     override fun onNewLocationMatcherResult(locationMatcherResult: LocationMatcherResult) {
+ *       mutableLocation.value = locationMatcherResult
+ *     }
+ *
+ *     override fun onNewRawLocation(rawLocation: Location) {
+ *       // no op
+ *     }
+ *   }
+ *
+ *   override fun onAttached(mapboxNavigation: MapboxNavigation) {
+ *     mapboxNavigation.registerLocationObserver(locationObserver)
+ *   }
+ *
+ *   override fun onDetached(mapboxNavigation: MapboxNavigation?) {
+ *     mapboxNavigation?.unregisterLocationObserver(locationObserver)
+ *   }
+ * }
+ * ```
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+interface MapboxNavigationObserver {
+    /**
+     * Emits the active MapboxNavigation object.
+     * After registering through [MapboxNavigationApp.registerObserver], the
+     * current [MapboxNavigation] object will be emitted.
+     */
+    fun onAttached(mapboxNavigation: MapboxNavigation)
+
+    /**
+     * Is called when all activities are destroyed and the car session is destroyed.
```

Comment:
```suggestion
     * Is called when all activities or services are destroyed/detached and there are no consumers of `MapboxNavigation`.
```

Reply:
But as mentioned above, this should be granular per each lifecycle in my opinion.

Reply:
Updated this to be more about `LifecycleOwner`. Services don't have a lifecycle, but you can still create your own LifecycleOwner (which is what androidx.car lib did). That way it is also granular.

I don't see harm in having the `attachAllActivities` available to developers. But maybe we should organize the comments or code to suggest using "attach/detach" functions.